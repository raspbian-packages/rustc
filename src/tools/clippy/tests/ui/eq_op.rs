#![feature(plugin)]
#![plugin(clippy)]

#[warn(eq_op)]
#[allow(identity_op, double_parens, many_single_char_names)]
#[allow(no_effect, unused_variables, unnecessary_operation, short_circuit_statement)]
#[warn(nonminimal_bool)]
fn main() {
    // simple values and comparisons
    1 == 1;
    "no" == "no";
    // even though I agree that no means no ;-)
    false != false;
    1.5 < 1.5;
    1u64 >= 1u64;

    // casts, methods, parentheses
    (1 as u64) & (1 as u64);
    1 ^ ((((((1))))));

    // unary and binary operators
    (-(2) < -(2));
    ((1 + 1) & (1 + 1) == (1 + 1) & (1 + 1));
    (1 * 2) + (3 * 4) == 1 * 2 + 3 * 4;

    // various other things
    ([1] != [1]);
    ((1, 2) != (1, 2));
    vec![1, 2, 3] == vec![1, 2, 3]; //no error yet, as we don't match macros

    // const folding
    1 + 1 == 2;
    1 - 1 == 0;

    1 - 1;
    1 / 1;
    true && true;

    true || true;


    let a: u32 = 0;
    let b: u32 = 0;

    a == b && b == a;
    a != b && b != a;
    a < b && b > a;
    a <= b && b >= a;

    let mut a = vec![1];
    a == a;
    2*a.len() == 2*a.len(); // ok, functions
    a.pop() == a.pop(); // ok, functions

    use std::ops::BitAnd;
    struct X(i32);
    impl BitAnd for X {
        type Output = X;
        fn bitand(self, rhs: X) -> X {
            X(self.0 & rhs.0)
        }
    }
    impl<'a> BitAnd<&'a X> for X {
        type Output = X;
        fn bitand(self, rhs: &'a X) -> X {
            X(self.0 & rhs.0)
        }
    }
    let x = X(1);
    let y = X(2);
    let z = x & &y;

    #[derive(Copy, Clone)]
    struct Y(i32);
    impl BitAnd for Y {
        type Output = Y;
        fn bitand(self, rhs: Y) -> Y {
            Y(self.0 & rhs.0)
        }
    }
    impl<'a> BitAnd<&'a Y> for Y {
        type Output = Y;
        fn bitand(self, rhs: &'a Y) -> Y {
            Y(self.0 & rhs.0)
        }
    }
    let x = Y(1);
    let y = Y(2);
    let z = x & &y;
}
