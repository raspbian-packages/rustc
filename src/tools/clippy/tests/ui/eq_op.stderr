error: this boolean expression can be simplified
  --> $DIR/eq_op.rs:37:5
   |
37 |     true && true;
   |     ^^^^^^^^^^^^ help: try: `true`
   |
   = note: `-D nonminimal-bool` implied by `-D warnings`

error: this boolean expression can be simplified
  --> $DIR/eq_op.rs:39:5
   |
39 |     true || true;
   |     ^^^^^^^^^^^^ help: try: `true`

error: this boolean expression can be simplified
  --> $DIR/eq_op.rs:45:5
   |
45 |     a == b && b == a;
   |     ^^^^^^^^^^^^^^^^ help: try: `a == b`

error: this boolean expression can be simplified
  --> $DIR/eq_op.rs:46:5
   |
46 |     a != b && b != a;
   |     ^^^^^^^^^^^^^^^^ help: try: `a != b`

error: this boolean expression can be simplified
  --> $DIR/eq_op.rs:47:5
   |
47 |     a < b && b > a;
   |     ^^^^^^^^^^^^^^ help: try: `a < b`

error: this boolean expression can be simplified
  --> $DIR/eq_op.rs:48:5
   |
48 |     a <= b && b >= a;
   |     ^^^^^^^^^^^^^^^^ help: try: `a <= b`

error: equal expressions as operands to `==`
  --> $DIR/eq_op.rs:10:5
   |
10 |     1 == 1;
   |     ^^^^^^
   |
   = note: `-D eq-op` implied by `-D warnings`

error: equal expressions as operands to `==`
  --> $DIR/eq_op.rs:11:5
   |
11 |     "no" == "no";
   |     ^^^^^^^^^^^^

error: equal expressions as operands to `!=`
  --> $DIR/eq_op.rs:13:5
   |
13 |     false != false;
   |     ^^^^^^^^^^^^^^

error: equal expressions as operands to `<`
  --> $DIR/eq_op.rs:14:5
   |
14 |     1.5 < 1.5;
   |     ^^^^^^^^^

error: equal expressions as operands to `>=`
  --> $DIR/eq_op.rs:15:5
   |
15 |     1u64 >= 1u64;
   |     ^^^^^^^^^^^^

error: equal expressions as operands to `&`
  --> $DIR/eq_op.rs:18:5
   |
18 |     (1 as u64) & (1 as u64);
   |     ^^^^^^^^^^^^^^^^^^^^^^^

error: equal expressions as operands to `^`
  --> $DIR/eq_op.rs:19:5
   |
19 |     1 ^ ((((((1))))));
   |     ^^^^^^^^^^^^^^^^^

error: equal expressions as operands to `<`
  --> $DIR/eq_op.rs:22:5
   |
22 |     (-(2) < -(2));
   |     ^^^^^^^^^^^^^

error: equal expressions as operands to `==`
  --> $DIR/eq_op.rs:23:5
   |
23 |     ((1 + 1) & (1 + 1) == (1 + 1) & (1 + 1));
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: equal expressions as operands to `&`
  --> $DIR/eq_op.rs:23:6
   |
23 |     ((1 + 1) & (1 + 1) == (1 + 1) & (1 + 1));
   |      ^^^^^^^^^^^^^^^^^

error: equal expressions as operands to `&`
  --> $DIR/eq_op.rs:23:27
   |
23 |     ((1 + 1) & (1 + 1) == (1 + 1) & (1 + 1));
   |                           ^^^^^^^^^^^^^^^^^

error: equal expressions as operands to `==`
  --> $DIR/eq_op.rs:24:5
   |
24 |     (1 * 2) + (3 * 4) == 1 * 2 + 3 * 4;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: equal expressions as operands to `!=`
  --> $DIR/eq_op.rs:27:5
   |
27 |     ([1] != [1]);
   |     ^^^^^^^^^^^^

error: equal expressions as operands to `!=`
  --> $DIR/eq_op.rs:28:5
   |
28 |     ((1, 2) != (1, 2));
   |     ^^^^^^^^^^^^^^^^^^

error: equal expressions as operands to `==`
  --> $DIR/eq_op.rs:32:5
   |
32 |     1 + 1 == 2;
   |     ^^^^^^^^^^

error: equal expressions as operands to `==`
  --> $DIR/eq_op.rs:33:5
   |
33 |     1 - 1 == 0;
   |     ^^^^^^^^^^

error: equal expressions as operands to `-`
  --> $DIR/eq_op.rs:33:5
   |
33 |     1 - 1 == 0;
   |     ^^^^^

error: equal expressions as operands to `-`
  --> $DIR/eq_op.rs:35:5
   |
35 |     1 - 1;
   |     ^^^^^

error: equal expressions as operands to `/`
  --> $DIR/eq_op.rs:36:5
   |
36 |     1 / 1;
   |     ^^^^^

error: equal expressions as operands to `&&`
  --> $DIR/eq_op.rs:37:5
   |
37 |     true && true;
   |     ^^^^^^^^^^^^

error: equal expressions as operands to `||`
  --> $DIR/eq_op.rs:39:5
   |
39 |     true || true;
   |     ^^^^^^^^^^^^

error: equal expressions as operands to `&&`
  --> $DIR/eq_op.rs:45:5
   |
45 |     a == b && b == a;
   |     ^^^^^^^^^^^^^^^^

error: equal expressions as operands to `&&`
  --> $DIR/eq_op.rs:46:5
   |
46 |     a != b && b != a;
   |     ^^^^^^^^^^^^^^^^

error: equal expressions as operands to `&&`
  --> $DIR/eq_op.rs:47:5
   |
47 |     a < b && b > a;
   |     ^^^^^^^^^^^^^^

error: equal expressions as operands to `&&`
  --> $DIR/eq_op.rs:48:5
   |
48 |     a <= b && b >= a;
   |     ^^^^^^^^^^^^^^^^

error: equal expressions as operands to `==`
  --> $DIR/eq_op.rs:51:5
   |
51 |     a == a;
   |     ^^^^^^

error: taken reference of right operand
  --> $DIR/eq_op.rs:89:13
   |
89 |     let z = x & &y;
   |             ^^^^--
   |                 |
   |                 help: use the right value directly: `y`
   |
   = note: `-D op-ref` implied by `-D warnings`

error: equal expressions as operands to `/`
  --> $DIR/eq_op.rs:97:20
   |
97 |     const D: u32 = A / A;
   |                    ^^^^^

error: aborting due to 34 previous errors

